samples/core/multi-device/main.c

```diff
@@ -0,0 +1,646 @@
+
+    // Create device buffers, from which we will create the subbuffers for the
+    // subdevices.
+    const size_t grid_midpoint = y_dim / 2;
+    const size_t pad_grid_midpoint = pad_y_dim / 2;
+
+    if (diag_opts.verbose)
+    {
+        printf("done.\nInitializing device-side storage...");
+        fflush(stdout);
+    }
+
+    OCLERROR_PAR(dev_input_grid =
+                     clCreateBuffer(context,
+                                    CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR
+                                        | CL_MEM_HOST_NO_ACCESS,
+                                    input_bytes, h_input_grid, &error),
+                 error, hmask);
+    OCLERROR_PAR(dev_output_grid =
+                     clCreateBuffer(context,
+                                    CL_MEM_WRITE_ONLY | CL_MEM_ALLOC_HOST_PTR
+                                        | CL_MEM_HOST_READ_ONLY,
+                                    output_bytes, NULL, &error),
+                 error, bufin);
+    OCLERROR_PAR(dev_mask =
+                     clCreateBuffer(context,
+                                    CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR
+                                        | CL_MEM_HOST_NO_ACCESS,
+                                    mask_bytes, h_mask, &error),
+                 error, bufout);
+
+    if (diag_opts.verbose)
+    {
+        printf("done.\nSetting up sub-devices...");
+        fflush(stdout);
+    }
+
+    // Set up subdevices for kernel execution.
+    const size_t half_input_bytes =
+        sizeof(cl_float) * pad_x_dim * (pad_grid_midpoint + 1);
+    const size_t input_offset =
+        sizeof(cl_float) * pad_x_dim * (pad_grid_midpoint - 1);
+    const size_t half_output_bytes = sizeof(cl_float) * x_dim * grid_midpoint;
+
+    cl_uint subdevice = 0;
+    for (; subdevice < subdev_count; ++subdevice)
+    {
+        // Create kernel.
+        if (diag_opts.verbose)
+        {
+            printf("\n  Creating kernel and command queue of sub-device %d...",
+                   subdevice);
+            fflush(stdout);
+        }
+
+        OCLERROR_PAR(convolutions[subdevice] =
+                         clCreateKernel(program, "convolution_3x3", &error),
+                     error, bufmask);
+
+        // Initialize queues for command execution on each device.
+#if CL_HPP_TARGET_OPENCL_VERSION >= 200
+        cl_command_queue_properties props[] = { CL_QUEUE_PROPERTIES,
+                                                CL_QUEUE_PROFILING_ENABLE, 0 };
+        OCLERROR_PAR(sub_queues[subdevice] = clCreateCommandQueueWithProperties(
+                         context, subdevices[subdevice], props, &error),
+                     error, conv);
+#else
+        OCLERROR_PAR(sub_queues[subdevice] = clCreateCommandQueue(
+                         context, subdevices[subdevice],
+                         CL_QUEUE_PROFILING_ENABLE, &error),
+                     error, conv);
+#endif
```

Comment:
This just follows the same approach as other examples, should it be removed then?

Reply:
IMHO yes, we should just use the 1.x version.  If we want to use the more complicated "with properties" version we'll need runtime checks to ensure it is supported - the compile-time checks aren't good enough.

If other examples have a similar pattern I'm fine leaving them as-is for now, but please file an issue so we don't forget about this.

Reply:
Yes, I see your point. I'll file an issue so this can be fixed in all the samples that implement this logic (namely: `core/binaries`, `core/ blur`, `core/reduce`,  `core/saxpy` and now `core/multi-device` C samples).

Reply:
https://github.com/KhronosGroup/OpenCL-SDK/issues/94